datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -391,48 +363,71 @@ impl CrossJoinStream {
+        match ready!(self.right.poll_next_unpin(cx)) {
+            None => Poll::Ready(None),
+            Some(Ok(right_batch)) => {
+                self.join_metrics.input_batches.add(1);
+                self.join_metrics.input_rows.add(right_batch.num_rows());
+
+                let join_timer = self.join_metrics.join_time.timer();
+                let result_batch =
+                    build_batch(self.schema.clone(), left_data, right_batch)?;
                 join_timer.done();
+
                 self.join_metrics.output_batches.add(1);
-                self.join_metrics.output_rows.add(batch.num_rows());
+                self.join_metrics.output_rows.add(result_batch.num_rows());
+
+                Poll::Ready(Some(Ok(result_batch)))
             }
-            self.left_index += 1;
-            return Poll::Ready(Some(result));
+            Some(Err(err)) => Poll::Ready(Some(Err(err))),
         }
-        self.left_index = 0;
-        self.right
-            .poll_next_unpin(cx)
-            .map(|maybe_batch| match maybe_batch {
-                Some(Ok(batch)) => {
-                    let join_timer = self.join_metrics.join_time.timer();
-                    let result =
-                        build_batch(self.left_index, &batch, left_data, &self.schema);
-                    self.join_metrics.input_batches.add(1);
-                    self.join_metrics.input_rows.add(batch.num_rows());
-                    if let Ok(ref batch) = result {
-                        join_timer.done();
-                        self.join_metrics.output_batches.add(1);
-                        self.join_metrics.output_rows.add(batch.num_rows());
-                    }
-                    self.left_index = 1;
-
-                    let mut right_batch = self.right_batch.lock();
-                    *right_batch = Some(batch);
-
-                    Some(result)
-                }
-                other => other,
+    }
+}
+
+/// Joins the left and right batches by iterating over right (probe) side one by one.
+/// Since the rows of right batch are processed in order, right order is preserved.
+fn build_batch(
+    schema: Arc<Schema>,
+    left_data: &RecordBatch,
+    right_batch: RecordBatch,
+) -> Result<RecordBatch> {
+    let mut result_batch = RecordBatch::new_empty(schema.clone());
+    let mut right_index = 0;
+    while right_index < right_batch.num_rows() {
+        let right_copies: Vec<Arc<dyn Array>> = right_batch
+            .columns()
+            .iter()
+            .map(|right_column| {
+                compute::take(
+                    right_column.as_ref(),
+                    &PrimitiveArray::<UInt64Type>::from_value(
+                        right_index as u64,
+                        left_data.num_rows(),
+                    ),
+                    None,
+                )
+                .map_err(|e| datafusion_common::DataFusionError::ArrowError(e, None))
             })
+            .collect::<Result<Vec<_>>>()?;
```

Comment:
We have a util for this work called `get_array_ref_at_indices` use that one instead.